### PR TITLE
[refactor] Simplify parameter mismatch error code

### PR DESCRIPTION
**Pull request by ntrel**
* Use one `OutBuffer` instead of 3, reducing memory allocations.
* Avoid repeating common string messages.
* Add a missing test for typesafe variadics.

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -4780,19 +4780,20 @@ extern (C++) final class TypeFunction : TypeNext
...
-        as.printf("`%s` of type `%s`", arg.toChars(), at);
-        OutBuffer ps;
-        ps.printf("`%s`", parameterToChars(par, this, qual));
         OutBuffer buf;
-        buf.printf(format, as.peekString(), ps.peekString());
+        // only mention rvalue if it's relevant
+        const rv = !arg.isLvalue() && par.storageClass & (STC.ref_ | STC.out_);
+        const char* fmt = rv ?
+            "cannot pass rvalue argument `%s` of type `%s` to parameter `%s`" :
+            "cannot pass argument `%s` of type `%s` to parameter `%s`";
```
Comment (jacob-carlborg) [inline src/dmd/mtype.d]: Seems a bit unnecessary to duplicate the whole format string when the only difference is `rvalue`. Perhaps `rvalue` can be passed as a parameter to `printf` as well?

**Review by RazvanN7 — APPROVED** [review]

File: src/dmd/mtype.d (hunk 1 above)
Comment (wilzbach) [reply src/dmd/mtype.d]: Yep, last time I tried something in this direction, it was complained that this increases the binary size.
It's just "cannot pass%s" and then `rv ? " rvalue" : ""`, no?

Comment (ntrel) [reply src/dmd/mtype.d]: Fix is here: https://github.com/ntrel/dmd/commit/6219e5e189ad7b758488fc751187ddbe6b792909

Comment (wilzbach) [reply src/dmd/mtype.d]: Cool. Would you mind submitting it as PR too? Thanks!

**Pull request merged**